### PR TITLE
Notebook testing

### PR DESCRIPTION
**Pull request by davidt0x**
This PR implements CI testing of brainiak and the example notebooks on Princeton research computing cluster della. I am still trying to work out some issues with failing or hanging tests on della. For now, I will have these tests skipped.

### CODE REVIEW

**Comment by mihaic** [comment]
@davidt0x, thank you for the hard work! Did the tests run on Jenkins?

**Comment by mihaic** [comment]
@davidt0x, I'm fixing some Myst errors today. Please don't change the notebooks in the meantime.

**Comment by mihaic** [comment]
@davidt0x, Myst-NB works now. Please have a look at the Conda NumPy error.

**Comment by davidt0x** [comment]
Thanks @mihaic, yeah, i am working on the numpy error in the conda build. This is a result of switching to the new build dependencies system in pyproject.toml. setup_requires is no longer specified but I think this is what meta.yaml is reading to setup the conda build environment dependencies. I think this is the issue. I am going to try to manually put them in meta.yaml as I don't know how to read them from pyproject.toml in Jinja.

File: setup.py [hunk 1]
```diff
@@ -115,57 +108,67 @@ def finalize_options(self):
...
+		"seaborn",
+		"holoviews",
+		"pyOpenSSL",
+		"awscli",
+		"bcrypt",
+		"indexed_gzip",
+		"inflect",
+		"ipython",
+		"jupyter",
+                "mypy @ git+https://github.com/python/mypy.git",
```
Comment (mihaic) [inline setup.py:148]: The mypy fix has been released:
https://github.com/python/mypy/issues/13385#issuecomment-1258454516
```suggestion
                "mypy",
```

**Comment by CameronTEllis** [comment]
@mihaic @davidt0x Is this PR waiting for review or is there still more to do?

**Comment by mihaic** [comment]
@CameronTEllis, the PR is ready for merging as soon as the tests pass. The only one missing now is the Conda build. Until we fix it, even if we extract the commit to fix fmrisim issue #520, we would likely still not be able to build the Conda packages.

**Comment by CameronTEllis** [comment]
@mihaic I will review it in a few days so that once the conda build is working the PR is ready to go.

File: pr-check.sh [hunk 2]
```diff
@@ -43,18 +71,32 @@ function remove_venv_venv {
...
+
 }
 
 function deactivate_conda_venv {
-    source deactivate
+
+    if [[ "$is_della" == true ]]; then
+        conda deactivate
+    else
+        source deactivate
```
Comment (CameronTEllis) [inline pr-check.sh:98]: I don't think that this is della specific, it is generally encouraged to use conda activate and conda deactivate

File: pr-check.sh [hunk 3]
```diff
@@ -18,6 +18,34 @@
 
 set -e
 
+# Check whether we are running on Princeton's della compute cluster.
+is_della=false
```
Comment (CameronTEllis) [inline pr-check.sh:22]: This script has a lot of carve outs for Princeton's set up, which isn't a versatile solution for other groups nor will it future proof Brainiak within Princeton (e.g., if Della changes). Is there a way instead to make a script that replaces this one with the della-specific commands that are needed?

File: pr-check.sh [hunk 4]
```diff
@@ -106,42 +148,80 @@ $activate_venv $venv || {
...
+    ./run-checks.sh || exit_with_error_and_venv "run-checks failed"
+fi
 
 # run tests
-./run-tests.sh $sdist_mode || \
-    exit_with_error_and_venv "run-tests failed"
+if [[ "$is_della" == true ]]; then
+    echo "Running on della head node, need to request time on a compute node"
+    export BRAINIAKDEV_MPI_COMMAND=srun
+    salloc -t 03:00:00 -N 1 -n 16 sh run-tests.sh $sdist_mode || \
```
Comment (CameronTEllis) [inline pr-check.sh:195]: Seems like this isn't della specific but just slurm specific

File: tests/test_notebooks.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+from testbook import testbook
+
+
+nb_files = glob.glob("docs/examples/**/*.ipynb", recursive=True)
+
+mpi_notebooks = ["htfa", "FCMA", "SRM"]
+
+nb_tests = []
+for f in nb_files:
```
Comment (CameronTEllis) [inline tests/test_notebooks.py:13]: Outputs here are specific to della but without clear reason why

**Review by CameronTEllis — COMMENTED** [review]
I have completed a review of 50/75 files, with the remaining files being outside of my expertise (e.g., .conda/build.sh, .github/workflows/main.yml, setup.py). Moreover, I did not review the docs/examples/ notebooks because I assume they are copies from aperture. If this is not the case, let me know.

My main concern is that mixed in with a lot of needed and important changes (e.g., changing type calls, using get_fdata) is a lot additions that are specific to Princeton's set up on Della. This seems like it could confuse users from other institutions and will lead to clutter in the code. I am not sure about the appropriate solution, but I wonder if it would be better to have a branch on brainiak that has these Della specific scripts.

**Comment by mihaic** [comment]
Thank you for your review, @CameronTEllis.

You are right, Princeton-specific code is an issue. When we decided to incorporate the Aperture notebooks for testing, we observed long test times with GitHub infrastructure. We identified Princeton infrastructure as an alternative that will provide fast test times. To improve clarity, I propose we document the use of Della at the beginning of `pr-check.sh`. Going forward, Della use will be separated, as started in `.github/workflows/della_notebooks.yml`.

**Comment by lcnature** [comment]
Hi I was facing some issues of package incompatibility and started what I realized now as redoing some of the work in the PR. Just wondering if there is anything I can help to make this PR move forward. Thanks!

**Comment by mihaic** [comment]
Thanks for your offer to help, @lcnature. The PR is looking good with all tests passing. @davidt0x is working hard on the Conda builds and we will merge the PR as soon they are fixed. Currently, Conda is taking hours to resolve the dependencies on Linux, but Mamba seems to make faster progress. If you feel like, please try building the MacOS Conda package.

**Comment by lcnature** [comment]
> Thanks for your offer to help, @lcnature. The PR is looking good with all tests passing. @davidt0x is working hard on the Conda builds and we will merge the PR as soon they are fixed. Currently, Conda is taking hours to resolve the dependencies on Linux, but Mamba seems to make faster progress. If you feel like, please try building the MacOS Conda package.

Thanks! Ah MacOS is the only thing that I cannot help much as I don't use Mac. Indeed conda is a pain when it comes to solving dependencies and mamba seems to be often the way to go.

**Comment by mihaic** [comment]
@vineetbansal, thank you very much for the fixes.

Although Codecov did not report the coverage results, they look fine on their website:
https://app.codecov.io/gh/brainiak/brainiak/pull/511

@davidt0x, thank you for seeing this huge PR through. It is finally time to merge it!

**Review by mihaic — APPROVED** [review]

**Pull request merged**